src/notification/notificationmodel.cpp

```diff
@@ -0,0 +1,114 @@
+#include <QDebug>
+#include <QTime>
+
+#include "notificationmodel.h"
+
+NotificationModel::NotificationModel(QObject* parent)
+    : QAbstractListModel(parent)
+{
+    // Add all roles in _roles
+    for(const auto key : _roleNames.keys()) {
+        _roles.append(key);
+    }
+}
+
+void NotificationModel::add(QVariant text, QVariant icon, QVariant color, QVariant severity)
+{
+    // Create a new row
+    const QString time = QTime::currentTime().toString(QStringLiteral("[hh:mm:ss]"));
+    const int line = rowCount();
+    beginInsertRows(QModelIndex(), line, line);
+    insertRows(line, 1);
+    setData(index(line), time, NotificationModel::Time);
+    setData(index(line), text, NotificationModel::Text);
+    setData(index(line), color, NotificationModel::Color);
+    setData(index(line), severity, NotificationModel::Severity);
+    setData(index(line), icon, NotificationModel::Icon);
+    endInsertRows();
+}
+
+bool NotificationModel::insertRows(int row, int count, const QModelIndex& parent)
+{
+
+    // Check for valid input
+    if (count < 1 || row < 0 || row > rowCount(parent)) {
+        return false;
+    }
+
+    // Add all request rows
+    while(count --> 0) {
+        for(auto* vector : _rows) {
```

Comment:
why do we use `auto*` here instead of just `auto`?

Reply:
To not get a copy of it, AFAIK the compiler will set auto to T.

Reply:
That's cool.